File: spawner.js
const {spawnChild} = require('./spawnutil');

async function main() {
  try {
    // Spawn two child processes
    const child1Promise = spawnChild('bash', ['sp1.sh', `${process.argv[2]}`]);
    const child2Promise = spawnChild('bash', ['sp2.sh']);

    // Wait for the first child process to exit
    const firstChild = await Promise.race([child1Promise, child2Promise]);

    console.log('First child to exit:', firstChild);

    // Wait for the remaining child process to exit
    const remainingChild = (firstChild === await child1Promise) 
        ? await child2Promise 
        : await child1Promise;

    console.log('Second child to exit:', remainingChild);
    process.exit(0);
} catch (error) {
    console.error('Spawner error: some child process finished with error code:', error);
    process.exit(12);
  }
}

(
    async () => await main()
)();

File: Spawner.java
import java.io.*;

public class Spawner {
    public static void main(String[] args) {
        try {
            // Command to execute the first bash script
            String[] command1 = {"bash", "sp1.sh", args[0]};
            // Command to execute the second bash script
            String[] command2 = {"bash", "sp2.sh"};

            // Create process builders for both commands
            ProcessBuilder processBuilder1 = new ProcessBuilder(command1);
            ProcessBuilder processBuilder2 = new ProcessBuilder(command2);

            // Inherit input, output, and error streams
            processBuilder1.inheritIO();
            processBuilder2.inheritIO();

            // Start both processes
            Process process1 = processBuilder1.start();
            Process process2 = processBuilder2.start();

            // Wait for both processes to complete
            int exitCode2 = process2.waitFor();
            int exitCode1 = process1.waitFor();

            if (exitCode2 == 0) {
              System.out.println("sp2 exited with code: " + exitCode2);
            } else { 
              System.exit(12);
            }
            if (exitCode1 == 0) {
              System.out.println("sp1 exited with code: " + exitCode1);
            } else { 
              System.exit(12);
            }

        } catch (IOException | InterruptedException e) {
            e.printStackTrace();
        }
    }
}
